Add cloud function to flag cryptominers in Mixpanel [BT-51]

breilly2 opened the pull request:
There is a module running somewhere in the leonardo part of the system that identifies cryptomining activity based on usage patterns. That module shuts down the runtime and notifies leonardo so that the leonardo database can be updated. That notification contains information about the runtime instance but nothing about who owns it. Leonardo uses the runtime info to resolve who owns the project and the runtime. Since they will have already resolved that, they have agreed to also send a Pub/Sub message with that information so that we can update Mixpanel.

To avoid having to create an authenticated endpoint in the existing bard GAE app, this PR adds a cloud function that will be triggered by Pub/Sub messages.

There is some copy/pasta code here because, while it makes sense to use the same GCP project that already has the Mixpanel token in Secret Manager, the GAE app and cloud function are distinct and separate deployments. They could share code, but the build/deploy process would need to be a little more sophisticated, such as copying the necessary source files into a scratch area from which to build/deploy.

I also found that Mixpanel seems to have updated their REST API. I could find no documentation for the old API so I had to use the new one, hence `fetchMixpanelV2`.

breilly2 committed:
Add cloud function to flag cryptominers in Mixpanel

breilly2 commented on cloud_functions/function.js:
```diff
@@ -0,0 +1,67 @@
+
+
+const flagCryptominer = async (message, context) => {
+  const subjectId = Buffer.from(message.data, 'base64').toString();
```
Hang on... forgot to change this to a JSON payload. Converting PR to draft.

breilly2 committed:
Changed message payload to JSON

kshakir reviewed (commented):
Should this code have tests?

panentheos commented on cloud_functions/function.js:
```diff
@@ -0,0 +1,72 @@
+
+const flagCryptominer = async (message, context) => {
+  const data = JSON.parse(Buffer.from(message.data, 'base64').toString())
+  const token = await getSecret({ project, secretName: 'mixpanel-api' })
```
TOL: Instead of doing this on every call, we could initialize it once in the global scope. However, that would make the secret 'stateful', and since this is probably not a high-frequency event, it might be worth keeping things simple instead of optimizing.

breilly2 replied:
I had planned to do this but forgot before creating the PR. Also wasn't sure about error handling, but I think I got that sorted out now.

panentheos commented on cloud_functions/function.js:
```diff
@@ -0,0 +1,72 @@
+      $token: token,
+      $distinct_id: `google:${data.googleSubjectId}`,
+      $set: {
+        'Is Cryptominer': true
```
Could this be `isCryptominer` instead? We can always create a translation in the lexicon if we want an English string in the UI.

panentheos commented on cloud_functions/function.js:
```diff
@@ -0,0 +1,72 @@
+  })
+}
+
+// TODO: Share this code with the bard GAE app (will probably require a more sophisticated build/deploy process)
```
Why weren't we able to share this code as-is?

breilly2 replied:
The directory structure and build scripts for both the GAE app and Cloud Function would need to change in order to make sure the right code went to the right deployments.

breilly2 committed:
Flag cryptominers, with tests (#24)

* Added some tests
Switched to Mixpanel's Node.js SDK

* Lots of refactoring.
Added tests for interactions with Mixpanel SDK.

* Feedback on tests and related code structure, including consolidation of error reporting.

breilly2 commented:
Tests added. Ready for re-review.

panentheos commented on cloud_functions/ValidationError.js:
```diff
@@ -0,0 +1,9 @@
+class ValidationError extends Error {
```
What's the rationale for creating this subclass? It doesn't look like we have any logic that makes decisions based on it.

breilly2 replied:
Ah. I had done something with this at one point, but it looks like it turns out to not be needed. I'll probably remove it (after double-checking that I didn't miss an intended usage).

panentheos commented on cloud_functions/flagCryptominer.js:
```diff
@@ -0,0 +1,35 @@
+const run = (mixpanel, message) => {
+  try {
+    const userSubjectId = validate(message)
+    if (userSubjectId) {
```
It looks like `validate` already ensures that we have a `userSubjectId`, so we probably don't need this condition.

breilly2 replied:
Good catch. This is left over from then `validate` just logged instead of throwing an exception.

panentheos commented on cloud_functions/flagCryptominer.js:
```diff
@@ -0,0 +1,35 @@
+    if (userSubjectId) {
+      mixpanel.people.set(`google:${userSubjectId}`, { 'isCryptominer': true }, err => { if (err) throw err })
+    }
+  } catch (error) {
```
TOL: Another way to factor this would be to move the `try`/`catch` into a wrapper, in order to fully separate out all the bits related to special error handling (which might make it easier to document). E.g.
```
const run = withErrorLogging((mixpanel, message) => {
  ...
})
```

breilly2 replied:
Ah, of course. Just like we do in terra-ui. I'll probably make that change. Thanks!

panentheos commented on cloud_functions/function.js:
```diff
@@ -0,0 +1,8 @@
+
+module.exports.flagCryptominer = async (message, context) => {
+  const mixpanel = await getMixpanel()
+  flagCryptominer.run(mixpanel, message)
```
Assuming this kicks off an async call, we should make sure we're waiting for it to finish before returning. Otherwise it's possible the instance could be shut down with unfinished requests in rare cases. (That might mean promisifying the mixpanel call further down if needed)

breilly2 replied:
I'm pretty sure I dug into this and found that it's not making an async call. o_O I'll dig again because it is very strange, and maybe worth a comment since it does look wrong.

panentheos replied:
I glanced at the docs and they're not super clear, but based on the current signature of `people.set`, it looks like it might be callback based. If that's the case, then we'd need to wrap the call in an explicit Promise to wire everything up.

breilly2 replied:
Ah, yes. Thanks for catching this! I do remember reading about callback-based APIs, but I guess I came to the wrong conclusion about what I needed to do. I'll fix this.

breilly2 replied:
After going through the exercise of writing it myself (which was a fun little exercise), I ended up using Node's `promisify` util function. Combined with the `withErrorReporting` suggestion, `flagCryptominer.js` now looks quite tidy! Thanks again for all the suggestions.

kshakir commented on cloud_functions/flagCryptominer.js:
```diff
@@ -0,0 +1,35 @@
+  let data
+  try {
+    data = message.data && JSON.parse(Buffer.from(message.data, 'base64').toString())
+  } catch (error) {
+    throw new ValidationError(error.message)
+  }
```
ToL: Must turn of my JVM spectacles. Rethrowing errors like this in the JVM would lose stack context, but from my first-google-hit this may be normal in JS land... https://stackoverflow.com/questions/25763837/how-to-specify-a-caused-by-in-a-javascript-error

breilly2 replied:
Indeed. I struggled with this myself.

As @panentheos pointed out, `ValidationError` may not even be needed. I had originally thought to use it to help isolate per-invocation errors (validation) from per-instance (i.e. Cloud Function instance, specifically fetching the Mixpanel token and initializing the Mixpanel SDK). However, that looks like it turned out to not be necessary so I'll probably remove it.

A happy consequence of that is that this `try/catch` will no longer be necessary, which also means `data` can be `const`. No more `let`!

kshakir commented on cloud_functions/test/flagCryptominer-error-tests.js:
```diff
@@ -0,0 +1,93 @@
+// These tests cover various error cases for the flagCryptominer cloud function. Since everything
+// about this operates in the background, we want to make sure that errors
+// Based on https://cloud.google.com/functions/docs/testing/test-background#unit_tests
```
"make sure that errors" ... trails off

breilly2 replied:
https://media.giphy.com/media/xB7FIRaxrtNaE/giphy.gif

Sorry. Got lost mid thought. Will fix.

breilly2 replied:
I think I got interrupted in the middle of rewriting this comment block. I completed that rewrite (rather than just completing the sentence). My aim is to comment first on what, then why, then implementation details.

panentheos commented on cloud_functions/utils.js:
```diff
@@ -0,0 +1,31 @@
+    const [version] = await client.accessSecretVersion({ name })
+    return version.payload.data.toString('utf8')
+  } catch (error) {
+    return null
```
Do we need to catch here, or can we just let the error propagate out?

breilly2 replied:
Interesting question. The context and the way this is used is a little different here in a cloud function than it is in GAE (and I wonder if it could be done differently there, too). Keeping the function exactly the same makes it easier to safely de-duplicate later. On the other hand, I could simplify this case and add to the comment above this, which might encourage improvement on the GAE side when de-duplicating. I think I'll give that a go.

kshakir reviewed (commented):
Doesn't look like Circle ran your tests? https://app.circleci.com/pipelines/github/DataBiosphere/bard/119/workflows/f092da17-9a97-4380-96df-14e69549b0ab/jobs/247

breilly2 committed:
PR feedback part 1

mcovarr commented on cloud_functions/test/flagCryptominer-error-tests.js:
```diff
@@ -0,0 +1,94 @@
+})
+
+test.serial('data is not JSON', async t => {
+  const mixpanel = utils.makeStubMixpanel()
```
can you make the mixpanel stub in the `beforeEach` to DRY things out?

breilly2 replied:
I think that would require `mixpanel` to be global and some kind of reset to be done in `afterEach`. That wouldn't be the end of the world, but I'd rather keep things out of global scope whenever possible.

I'd like for `console` to be the only exception causing these to have to run in serial rather than allowing the fact that these have to run in serial to justify adding more global state.

mcovarr commented on cloud_functions/test/flagCryptominer-error-tests.js:
```diff
@@ -0,0 +1,94 @@
+    data: Buffer.from('invalid').toString('base64')
+  }
+
+  await flagCryptominer.run(mixpanel, message)
```
likewise a function to wrap this `await`ed call?

breilly2 replied:
`flagCryptominer.run` is the unit of functionality under test, so I don't want to hide its use behind a wrapper function.

My philosophy for test code is that DRY-er code reduces readability of individual tests. It also increases the chance of making a change for the sake of one test that invalidates the original intent of other tests. In the worst case, the tests continue to pass but no longer test the thing they originally tested (I've seen this first-hand). When this gets noticed (likely later, likely by someone new to the code), it's a coin toss whether the intent is rediscovered and the tests fixed or the tests are just deleted because they appear useless and/or redundant.

panentheos replied:
I agree with @breilly2 here. The similarities seem more incidental than intentional, so I think keeping them independent is a good idea.

breilly2 committed:
PR feedback part 2

breilly2 committed:
Trying to run tests in CircleCI

breilly2 commented:
@kshakir Tests now run for PRs. I'd like to make the build-cloud-function check required, but I don't have admin access for this repo. Maybe @panentheos does or @petesantos does?

breilly2 committed:
Don't bother with yarn cache for now

petesantos commented:
> @kshakir Tests now run for PRs. I'd like to make the build-cloud-function check required, but I don't have admin access for this repo. Maybe @panentheos does or @petesantos does?

I added you as an admin and made the build-cloud-function check required.

panentheos commented on cloud_functions/flagCryptominer.js:
```diff
@@ -0,0 +1,24 @@
+
+const run = utils.withErrorReporting((mixpanel, message) => {
+  const userSubjectId = validate(message)
+  return promisify(mixpanel.people.set)(`google:${userSubjectId}`, { 'isCryptominer': true })
```
Nice use of `promisify`. Note that this is ok to use because the mixpanel library doesn't rely on `this` internally, so wrapping the bare function works fine.

breilly2 replied:
Thanks for pointing that out. I didn't see any mention of that in the node docs. 😐  Can we be sure that the mixpanel library doesn't rely on `this`? And can we be sure that that will always be true? I'm wondering if it would be better to write it out longhand in this case.

panentheos replied:
I checked the source of the library to verify it. I think it's unlikely to change, but if they did change it this would break, so if we're worried about that we should write it longhand. (Note that the current tests would not catch that, since they all stub out the mixpanel library)

breilly2 replied:
Okay. I'm less enamored of `promisify` now, but I'll leave it because risk seems low and I trust that failures will not be silent.

panentheos reviewed (approved):
Looks good. Thanks for iterating on this!
